benches/fraud.rs

```diff
@@ -0,0 +1,399 @@
+    category: u32,
+    amt: F64,
+    state: u32,
+    job: u32,
+    unix_time: i32,
+    city_pop: u32,
+    merchant: u32,
+    is_fraud: u32,
+}
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Deserialize, SizeOf)]
+struct Demographics {
+    cc_num: F64,
+    first: u32,
+    gender: u32,
+    street: u32,
+    city: u32,
+    state: u32,
+    zip: u32,
+    lat: F64,
+    long: F64,
+    city_pop: u32,
+    job: u32,
+    #[size_of(skip)]
+    dob: Date,
+}
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Deserialize, SizeOf)]
+struct Transaction {
+    #[size_of(skip)]
+    #[serde(deserialize_with = "primitive_date_time_from_str")]
+    trans_date_trans_time: PrimitiveDateTime,
+    cc_num: F64,
+    merchant: u32,
+    category: u32,
+    amt: F64,
+    trans_num: String,
+    unix_time: i32,
+    merch_lat: F64,
+    merch_long: F64,
+    is_fraud: u32,
+}
+
+fn primitive_date_time_from_str<'de, D: Deserializer<'de>>(
+    d: D,
+) -> Result<PrimitiveDateTime, D::Error> {
+    let s: String = Deserialize::deserialize(d)?;
+
+    let format = time::macros::format_description!("[year]-[month]-[day] [hour]:[minute]:[second]");
+
+    match time::PrimitiveDateTime::parse(&s, &format) {
+        Ok(o) => Ok(o),
+        Err(err) => Err(D::Error::custom(err)),
+    }
+}
+
+#[derive(Debug, Clone, Parser)]
+struct Args {
+    #[clap(long)]
+    workers: usize,
+
+    #[clap(long, default_value = DEFAULT_BATCH_SIZE)]
+    batch_size: usize,
+
+    #[clap(long, default_value = "benches/fraud_data/demographics.csv")]
+    demographics: String,
+
+    #[clap(long, default_value = "benches/fraud_data/transactions.csv")]
+    transactions: String,
+
+    #[clap(long)]
+    stdin: bool,
+
+    #[doc(hidden)]
+    #[clap(long = "bench", hide = true)]
+    __bench: bool,
+}
+
+type Weight = i32;
+
+struct Fraud {
+    dbsp: DBSPHandle,
+    hdemographics: CollectionHandle<Demographics, Weight>,
+    htransactions: CollectionHandle<Transaction, Weight>,
+}
+
+impl Fraud {
+    fn new(workers: usize) -> Self {
+        let (dbsp, (hdemographics, htransactions)) = Runtime::init_circuit(workers, |circuit| {
+            let (demographics, hdemographics) = circuit.add_input_zset::<Demographics, Weight>();
+            let (transactions, htransactions) = circuit.add_input_zset::<Transaction, Weight>();
+
+            let amounts = transactions.map_index(|t| {
+                let timestamp = t.trans_date_trans_time.assume_utc().unix_timestamp();
+                (t.cc_num, (timestamp, t.amt))
+            });
+
+            let transactions_by_ccnum = transactions.map_index(|t| (t.cc_num, t.clone()));
+            let demographics_by_ccnum = demographics.map_index(|d| (d.cc_num, d.clone()));
+
+            let enriched_transactions: Stream<
+                _,
+                OrdIndexedZSet<(F64, i64), (Transaction, Demographics), Weight>,
+            > = transactions_by_ccnum.join_index::<(), _, _, _, _, _>(
```

Comment:
why does join_index need a bound with almost everything _?

Reply:
Most of its type arguments are automatically inferred, except the logical timestamp. There are todo's in the code to refactor things so that this type is derived from the circuit.  Once that is done, `join` and variants won't need explicit type arguments.